nuxeo-runtime/nuxeo-runtime-cluster/src/main/java/org/nuxeo/runtime/cluster/ClusterServiceImpl.java

```diff
@@ -117,4 +123,148 @@ public void setNodeId(String nodeId) {
+         *           Here we don't want a new thread or a risk of deadlock, so we just commit the original transaction.
+         */
+        protected void runInSeparateTransaction(Runnable runnable) {
+            // check if there is a current transaction, before committing it
+            boolean transaction = TransactionHelper.isTransactionActiveOrMarkedRollback();
+            if (transaction) {
+                TransactionHelper.commitOrRollbackTransaction();
+            }
+            boolean completedAbruptly = true;
+            try {
+                if (transaction) {
+                    TransactionHelper.runInTransaction(runnable);
+                } else {
+                    runnable.run();
+                }
+                completedAbruptly = false;
+            } finally {
+                if (transaction) {
+                    // restart a transaction if there was one originally
+                    try {
+                        TransactionHelper.startTransaction();
+                    } finally {
+                        if (completedAbruptly) {
+                            // mark rollback-only if there was an exception
+                            TransactionHelper.setTransactionRollbackOnly();
+                        }
+                    }
+                }
+            }
+        }
+
+        /**
+         * Runs a {@link Runnable} atomically, in a cluster-wide critical section.
+         */
+        protected void runAtomicallyInternal(String key, Runnable runnable) {
+            String lockInfo = tryLock(key);
+            if (lockInfo != null) {
+                try {
+                    runnable.run();
+                } finally {
+                    unLock(key, lockInfo);
+                }
+            } else {
+                throw new RuntimeServiceException("Failed to acquire lock '" + key + "' after " + duration.toSeconds()
+                        + "s, owner: " + getLock(key));
+            }
+        }
+
+        // try to acquire the lock and fail if it takes too long
+        protected String tryLock(String key) {
+            log.debug("Trying to lock '{}'", key);
+            long deadline = System.nanoTime() + duration.toNanos();
+            long ttl = duration.multipliedBy(TTL_MULTIPLIER).toSeconds();
+            do {
+                // try to acquire the lock
+                String lockInfo = "node=" + nodeId + " time=" + Instant.now();
+                if (kvStore.compareAndSet(key, null, lockInfo, ttl)) {
+                    // lock acquired
+                    log.debug("Lock '{}' acquired after {}ms", () -> key,
+                            () -> (System.nanoTime() - (deadline - duration.toNanos())) / 1_000_000);
+                    return lockInfo;
+                }
+                // wait a bit before retrying
+                log.debug("  Sleeping on busy lock '{}' for {}ms", () -> key, () -> pollDelay.toMillis());
+                try {
```

Comment:
can be replaced by: ` log.debug("Sleeping on busy lock '{}' for {}ms", () -> key, pollDelay::toMillis);`